lib/SPIRVProducerPass.cpp

```diff
@@ -331,6 +331,14 @@ struct SPIRVProducerPass final : public ModulePass {
   // that |Ty| and its subtypes will need a corresponding SPIR-V type.
   void FindType(Type *Ty);
 
+  // Returns an equivalent type to |type|.
+  //
+  // By default, clspv maps both __constant and __global address space pointers
+  // to StorageBuffer storage class. In order to prevent duplicate types from
+  // being generated, clspv checks if the equivalent type has already been
+  // generated.
+  Type *EquivalentType(Type *type);
```

Comment:
The latest commit switches this around to actually use a canonical type. I think the previous version may have had problems with structs when there were pointer members, but I don't think those cases can be tested properly since they lead to invalid SPIR-V. Let me know what you think about this as an alternative.